### PR TITLE
Briefcase should work behind a proxy network

### LINKED ISSUE #86: Can/Should briefcase be usable behind a proxy
I'm trying to connect to http://opendatakit.appspot.com/ behind a proxy server. Getting following stack-trace. I suspect this is due to the proxy. 

Is briefcase suppose to work behind a proxy?

```
org.apache.http.conn.HttpHostConnectException: Connect to opendatakit.appspot.com:80 [opendatakit.appspot.com/74.125.138.153, opendatakit.appspot.com/2607:f8b0:4002:c0c:0:0:0:99] failed: Network is unreachable (connect failed)
        at org.apache.http.impl.conn.DefaultHttpClientConnectionOperator.connect(DefaultHttpClientConnectionOperator.java:158)
        at org.apache.http.impl.conn.PoolingHttpClientConnectionManager.connect(PoolingHttpClientConnectionManager.java:353)
        at org.apache.http.impl.execchain.MainClientExec.establishRoute(MainClientExec.java:380)
        at org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:236)
        at org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:184)
        at org.apache.http.impl.execchain.RetryExec.execute(RetryExec.java:88)
        at org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:110)
        at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:184)
        at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:82)
        at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:55)
        at org.opendatakit.briefcase.util.AggregateUtils.testServerConnectionWithHeadRequest(AggregateUtils.java:554)
        at org.opendatakit.briefcase.util.ServerUploader.testServerUploadConnection(ServerUploader.java:391)
        at org.opendatakit.briefcase.util.ServerConnectionTest.run(ServerConnectionTest.java:41)
        at org.opendatakit.briefcase.ui.ServerConnectionDialog.actionPerformed(ServerConnectionDialog.java:204)
        at javax.swing.AbstractButton.fireActionPerformed(AbstractButton.java:2022)
        at javax.swing.AbstractButton$Handler.actionPerformed(AbstractButton.java:2348)
        at javax.swing.DefaultButtonModel.fireActionPerformed(DefaultButtonModel.java:402)
        at javax.swing.DefaultButtonModel.setPressed(DefaultButtonModel.java:259)
        at javax.swing.plaf.basic.BasicButtonListener.mouseReleased(BasicButtonListener.java:252)
        at java.awt.Component.processMouseEvent(Component.java:6533)
        at javax.swing.JComponent.processMouseEvent(JComponent.java:3324)
        at java.awt.Component.processEvent(Component.java:6298)
        at java.awt.Container.processEvent(Container.java:2236)
        at java.awt.Component.dispatchEventImpl(Component.java:4889)
        at java.awt.Container.dispatchEventImpl(Container.java:2294)
        at java.awt.Component.dispatchEvent(Component.java:4711)
        at java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4888)
        at java.awt.LightweightDispatcher.processMouseEvent(Container.java:4525)
        at java.awt.LightweightDispatcher.dispatchEvent(Container.java:4466)
        at java.awt.Container.dispatchEventImpl(Container.java:2280)
        at java.awt.Window.dispatchEventImpl(Window.java:2746)
        at java.awt.Component.dispatchEvent(Component.java:4711)
        at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:758)
        at java.awt.EventQueue.access$500(EventQueue.java:97)
        at java.awt.EventQueue$3.run(EventQueue.java:709)
        at java.awt.EventQueue$3.run(EventQueue.java:703)
        at java.security.AccessController.doPrivileged(Native Method)
        at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:80)
        at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:90)
        at java.awt.EventQueue$4.run(EventQueue.java:731)
        at java.awt.EventQueue$4.run(EventQueue.java:729)
        at java.security.AccessController.doPrivileged(Native Method)
        at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:80)
        at java.awt.EventQueue.dispatchEvent(EventQueue.java:728)
        at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:201)
        at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
        at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:109)
        at java.awt.WaitDispatchSupport$2.run(WaitDispatchSupport.java:184)
        at java.awt.WaitDispatchSupport$4.run(WaitDispatchSupport.java:229)
        at java.awt.WaitDispatchSupport$4.run(WaitDispatchSupport.java:227)
        at java.security.AccessController.doPrivileged(Native Method)
        at java.awt.WaitDispatchSupport.enter(WaitDispatchSupport.java:227)
        at java.awt.Dialog.show(Dialog.java:1084)
        at java.awt.Component.show(Component.java:1671)
        at java.awt.Component.setVisible(Component.java:1623)
        at java.awt.Window.setVisible(Window.java:1014)
        at java.awt.Dialog.setVisible(Dialog.java:1005)
        at org.opendatakit.briefcase.ui.PushTransferPanel$DestinationActionListener.actionPerformed(PushTransferPanel.java:132)
        at javax.swing.AbstractButton.fireActionPerformed(AbstractButton.java:2022)
        at javax.swing.AbstractButton$Handler.actionPerformed(AbstractButton.java:2348)
        at javax.swing.DefaultButtonModel.fireActionPerformed(DefaultButtonModel.java:402)
        at javax.swing.DefaultButtonModel.setPressed(DefaultButtonModel.java:259)
        at javax.swing.plaf.basic.BasicButtonListener.mouseReleased(BasicButtonListener.java:252)
        at java.awt.Component.processMouseEvent(Component.java:6533)
        at javax.swing.JComponent.processMouseEvent(JComponent.java:3324)
        at java.awt.Component.processEvent(Component.java:6298)
        at java.awt.Container.processEvent(Container.java:2236)
        at java.awt.Component.dispatchEventImpl(Component.java:4889)
        at java.awt.Container.dispatchEventImpl(Container.java:2294)
        at java.awt.Component.dispatchEvent(Component.java:4711)
        at java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4888)
        at java.awt.LightweightDispatcher.processMouseEvent(Container.java:4525)
        at java.awt.LightweightDispatcher.dispatchEvent(Container.java:4466)
        at java.awt.Container.dispatchEventImpl(Container.java:2280)
        at java.awt.Window.dispatchEventImpl(Window.java:2746)
        at java.awt.Component.dispatchEvent(Component.java:4711)
        at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:758)
        at java.awt.EventQueue.access$500(EventQueue.java:97)
        at java.awt.EventQueue$3.run(EventQueue.java:709)
        at java.awt.EventQueue$3.run(EventQueue.java:703)
        at java.security.AccessController.doPrivileged(Native Method)
        at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:80)
        at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:90)
        at java.awt.EventQueue$4.run(EventQueue.java:731)
        at java.awt.EventQueue$4.run(EventQueue.java:729)
        at java.security.AccessController.doPrivileged(Native Method)
        at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:80)
        at java.awt.EventQueue.dispatchEvent(EventQueue.java:728)
        at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:201)
        at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
        at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
        at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
        at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
        at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
Caused by: java.net.ConnectException: Network is unreachable (connect failed)
        at java.net.PlainSocketImpl.socketConnect(Native Method)
        at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
        at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
        at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
        at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
        at java.net.Socket.connect(Socket.java:589)
        at org.apache.http.conn.socket.PlainConnectionSocketFactory.connectSocket(PlainConnectionSocketFactory.java:74)
        at org.apache.http.impl.conn.DefaultHttpClientConnectionOperator.connect(DefaultHttpClientConnectionOperator.java:141)
```

### PR DESCRIPTION
**Pull request by rclakmal**
Fixes #86 

**How to Test**
1. If you work behind a proxy try to pull from the aggregate and see whether you see the updated error message as follows. 
Updated the error message with a hint about proxy. 
proxywarn

2. Go to the Settings tab add your proxy settings and save. 
New Settings Tab
bbb4e8bc-1faf-11e7-8c68-19a895798480

3. Re-do your pulling. 
4. Minor test case: Close briefcase and restart to see whether your proxy settings are still saved. 

Also attached the jar file for quick testing purposes. 
[BriefCase_Proxy_Changes.zip](https://github.com/opendatakit/briefcase/files/917259/BriefCase_Proxy_Changes.zip)

### CODE REVIEW

**Comment by shivam-tripathi** [comment]
Works for me on Ubuntu 16.04 LTS

**Comment by yanokwa** [comment]
Thanks @rclakmal and @shivam-tripathi! I can confirm this works as expected, so now I can go through a code review.

Testing this on macOS is a little tricky because Java apps don't go through the normal proxying that macOS supports. What I did was:
1. Setup [tinyproxy](https://tinyproxy.github.io/) on an external server, allow traffic from my IP, and open up port 8888 on the server.
1. Use [LittleSnitch](https://www.obdev.at/products/littlesnitch/index.html) to block all outgoing traffic to everything but the external server.

Once that's done, Briefcase will error out on non-proxy connections and will work on proxied connections.

**Comment by rclakmal** [comment]
Thank you @shivam-tripathi and @yanokwa for your testings.  I myself tested on windows 7 and Linux Mint.

**Comment by icemc** [comment]
Works for me on windows

**Comment by rclakmal** [comment]
Thanks @icemc for your testing. :-)

File: src/org/opendatakit/briefcase/model/BriefcasePreferences.java [hunk 1]
```diff
@@ -143,4 +156,17 @@ public Preferences preferenceFactory(Class<?> node) {
     private static final BriefcasePreferences APPLICATION_SCOPED =
         new BriefcasePreferences(BriefcasePreferences.class, PreferenceScope.APPLICATION);
   }
+  
+  public static String getBriefCaseProxyType() {
```
Comment (daveycrockett) [inline src/org/opendatakit/briefcase/model/BriefcasePreferences.java]: For the sake of a symmetric interface, could we remove the below three static methods and refactor into one method, `public static ProxyConnection getBriefcaseProxyProperty()`?  Also the naming here feels a little smurfed to me (but I don't feel strongly).  I think "briefcase directory" makes sense, but these could probably just be "getProxyProperty" and "setProxyProperty."

File: src/org/opendatakit/briefcase/ui/SettingsPanel.java [hunk 2]
```diff
@@ -38,38 +56,102 @@ public SettingsPanel(MainBriefcaseWindow parentWindow) {
...
-                        .createSequentialGroup()
-                        .addContainerGap()
-                        .addGroup(
-                                groupLayout
-                                        .createParallelGroup(GroupLayout.Alignment.BASELINE)
-                                        .addComponent(txtBriefcaseDir, GroupLayout.PREFERRED_SIZE,
-                                                GroupLayout.DEFAULT_SIZE, GroupLayout.PREFERRED_SIZE)
-                                        .addComponent(btnChoose).addComponent(lblBriefcaseDirectory))
-                        .addContainerGap()));
+        groupLayout.setHorizontalGroup(
```
Comment (daveycrockett) [inline src/org/opendatakit/briefcase/ui/SettingsPanel.java]: Really like how the layout looks here!

File: src/org/opendatakit/briefcase/model/BriefcasePreferences.java (hunk 1 above)
Comment (daveycrockett) [inline src/org/opendatakit/briefcase/model/BriefcasePreferences.java]: It also might be cleaner/more efficient to have getProxyProperty return 'null' in the case of no proxy, rather than pass around a model with all empty strings in this case.  See related comments below.

File: src/org/opendatakit/briefcase/ui/SettingsPanel.java [hunk 3]
```diff
@@ -38,38 +56,102 @@ public SettingsPanel(MainBriefcaseWindow parentWindow) {
...
         navOrder.add(lblBriefcaseDirectory);
         navOrder.add(txtBriefcaseDir);
         navOrder.add(btnChoose);
+        
+        setCurrentProxySettings();
     }
+    
+    private void setCurrentProxySettings() {
+      String currentProxyType = BriefcasePreferences.getBriefCaseProxyType();
+	  if (currentProxyType.equals(ProxyConnection.ProxyType.NO_PROXY.toString())){
```
Comment (daveycrockett) [inline src/org/opendatakit/briefcase/ui/SettingsPanel.java]: null check here instead for no proxy, return immediately after setting the schemaComboBox?

File: src/org/opendatakit/briefcase/ui/SettingsPanel.java [hunk 4]
```diff
@@ -88,6 +170,42 @@ public void actionPerformed(ActionEvent e) {
...
+
+    }
+    
+    class ProxyConnectionCreateListener implements ActionListener {
+
+        @Override
+        public void actionPerformed(ActionEvent e) {
+        	ProxyConnection.ProxyType proxyType;
+            if (schemaComboBox.getSelectedIndex() == 0) {
+            	proxyType = ProxyConnection.ProxyType.NO_PROXY;
```
Comment (daveycrockett) [inline src/org/opendatakit/briefcase/ui/SettingsPanel.java]: Set proxy to null here and return immediately?

**Comment by rclakmal** [comment]
@daveycrockett Thanks for your useful review. I liked the changes you suggested. It's more clearer that way. I pushed the new changes. May be you could have a re-look :-)

File: src/org/opendatakit/briefcase/model/ProxyConnection.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+/**
+ * This class models a proxy connection.
+ * 
+ * @author rclakmal@gmail.com
+ *
+ */
+public class ProxyConnection {
+	private String host;
+	private Integer port;
+	private String proxyType;
```
Comment (daveycrockett) [inline src/org/opendatakit/briefcase/model/ProxyConnection.java]: Just curious, why the change to String for proxyType in this most recent commit?  I don't have a strong preference between String or enum, but if we're leaving it String can the below enum go away?  I have a feeling I'm missing something here.

**Comment by daveycrockett** [comment]
I only had one more comment about the most recent commit, but otherwise this looks great to me!  Thanks for making the changes.

File: src/org/opendatakit/briefcase/model/ProxyConnection.java (hunk 5 above)
Comment (rclakmal) [inline src/org/opendatakit/briefcase/model/ProxyConnection.java]: I thought about removing the enum too. But then I thought it is much clearer/extendable this way to maintain the proxyType without adding string constants to an external file.

Comment (daveycrockett) [inline src/org/opendatakit/briefcase/model/ProxyConnection.java]: I agree, the enum is explicit and adds clarity!  Is there any reason that an enum value can't be passed to the constructor for ProxyConnection, e.g. `public ProxyConnection(String host, Integer port, ProxyType proxyType) {` instead of `public ProxyConnection(String host, Integer port, String proxyType) {`?  This way it enforces that the caller has to specify a valid value of proxyType, rather than accepting any String. Might be overly defensive programming, but I think it might be applicable here.

Comment (rclakmal) [inline src/org/opendatakit/briefcase/model/ProxyConnection.java]: you are right...Well the doubt I had was you can't store anything other than String in `Preferences`. But I presume, I could have just used `Enum.valueOf(String)` to convert the String to the Enum type. 

I'll update PR in a while. Thanks :-)

**Comment by rclakmal** [comment]
@daveycrockett Adapted to use the enum. :-) Thanks for your review.

**Comment by lognaturel** [comment]
Nice work, @rclakmal! 

I wonder whether we should start thinking about the design of an ideal settings screen to shape this addition to it. It feels a little bit odd to have the "Save" button for proxy settings right next to the settings themselves rather than having a single opportunity to save/cancel at the bottom of the screen. Or perhaps the settings should just autosave on change in focus?

It appears that the Host and Port settings are disabled when I first access the settings screen even when the schema is set to http/https. I can enable them by switching the schema and then switching it back to again. Also, the fields are disabled after I save the proxy settings. I think they should probably always be enabled unless No Proxy is selected as the schema.

**Comment by yanokwa** [comment]
@rclakmal In other Java apps and in Firefox, I don't see an option for HTTPS proxies. My quick research suggests it's not a thing (http://stackoverflow.com/a/1512268/152938). Do you have a way of testing an HTTPS proxy? If no, we should remove it. Removing it would also simplify the UI because you wouldn't need the schema dropdown.

**Comment by daveycrockett** [comment]
:+1: good point @yanokwa re: https proxy not being a thing (why doesn't this immediately throw an IllegalArgumentException in DefaultProxyRoutePlanner's constructor?!).  Re: designing the settings screen @lognaturel, is this worth writing up in a separate issue?

**Comment by rclakmal** [comment]
Gents, 
Bit busy at the moment hunting for an internship. I will take this up again as soon as I can. Thank you for the tips and feedback :-)

**Comment by daveycrockett** [comment]
Hi @rclakmal, given all the back-and-forth you've already done with this, alternatively I'd be happy to fork what you have so far and add one more commit to make these last minor tweaks...let me know!

**Comment by rclakmal** [comment]
@daveycrockett Pls go ahead :)

**Comment by yanokwa** [comment]
I'm closing this so we can continue it at #124.
